src/Clash/Cores/Ethernet/Examples/FullUdpStack.hs

```diff
@@ -0,0 +1,282 @@
+
+Let's first implement `arpIcmpUdpStackC`:
+
+>>> :{
+arpIcmpUdpStackC
+  :: forall (dataWidth :: Nat) (dom :: Domain)
+   . HiddenClockResetEnable dom
+  => KnownNat dataWidth
+  => 1 <= dataWidth
+  => 1 <= DomainPeriod dom
+  => DomainPeriod dom <= 5 * 10^11
+  => KnownNat (DomainPeriod dom)
+  => Signal dom MacAddress
+  -- ^ My MAC Address
+  -> Signal dom (IPv4Address, IPv4Address)
+  -- ^ My IP address and the subnet
+  -> Circuit (PacketStream dom dataWidth (IPv4Address, UdpHeaderLite)) (PacketStream dom dataWidth (IPv4Address, UdpHeaderLite))
+  -- ^ UDP handler circuit
+  -> Circuit (PacketStream dom dataWidth EthernetHeader) (PacketStream dom dataWidth EthernetHeader)
+arpIcmpUdpStackC macAddressS ipS udpCkt = circuit $ \ethIn -> do
+  [arpEthIn, ipEthIn] <- packetDispatcherC (routeBy _etherType $ 0x0806 :> 0x0800 :> Nil) -< ethIn
+  ipTx <- ipLitePacketizerC <| packetBufferC d10 d4 <| icmpUdpStack <| packetBufferC d10 d4 <| filterMetaS (isForMyIp <$> ipS) <| ipDepacketizerLiteC -< ipEthIn
+  (ipEthOut, arpLookup) <- toEthernetStreamC macAddressS -< ipTx
+  arpEthOut <- arpC d10 d5 macAddressS (fst <$> ipS) -< (arpEthIn, arpLookup)
+  packetArbiterC RoundRobin -< [arpEthOut, ipEthOut]
+  where
+    icmpUdpStack = circuit $ \ipIn -> do
+      [icmpIn, udpIn] <- packetDispatcherC (routeBy _ipv4lProtocol $ 0x0001 :> 0x0011 :> Nil) -< ipIn
+      icmpOut <- icmpEchoResponderC @dom @dataWidth (fst <$> ipS) -< icmpIn
+      udpInParsed <- udpDepacketizerC -< udpIn
+      udpOutParsed <- udpPacketizerC (fst <$> ipS) <| udpCkt -< udpInParsed
+      packetArbiterC RoundRobin -< [icmpOut, udpOutParsed]
+    isForMyIp (ip, subnet) (_ipv4lDestination -> to) = to == ip || to == ipv4Broadcast ip subnet
+}:
+
+`arpIcmpUdpStackC` takes in:
+
+* a `Signal` `MacAddress`, our MAC address. `MacAddress` is simply a wrapper type around a @Vec @6 (BitVector 8)`. 
+
+* a signal that represents a pair of my IP address and the mask of the subnet (the subnet's prefix bits are set to 1). Like with `MacAddress`, `IPAddress` is a wrapper type around a ``Vec 4 (BitVector 8)`.
+
+* finally, a circuit that handles UPD packets. It has type `Circuit` (`PacketStream` `dom` `dataWidth` (`IPv4Address`, `UdpHeaderLite`)) `(PacketStream dom dataWidth (IPv4Address, UdpHeaderLite))`
+
+Line by line, the stack's implementation does the following:
+
+1. `packetDispatcherC` splits the `ethIn` into separate streams; this circuit takes a vector of predicates `Vec p (a -> Bool)`. It returns a vector `Vec p (PacketStream dom n a)`, and sends an incoming packet to the first matching entry.
+   We use the function `routeBy`, which makes a common design pattern much simpler to implement, namely we pattern match on the value of one of the fields of the `_meta` field: ARP packets have `_etherType` 0x0806, `_etherType` 0x0800. 
+   See the type of `routeBy` for the exact type. This splits the stream up into a vector of two signals: `[arpEthIn, ipEthIn]`.
```

Comment:
It doesn't really split it into signals. circuit-notation calls it ports. I think a general note on circuit-notation might be nice, because this syntax requires enabling the plugin. Maybe add a link to `Protocols.Plugin`?

Reply:
yes, linked plugin, changed to ports (with a short notice on the fact that ports have a forward and backward in `ArpLookup`)